sycl/test/basic_tests/group_async_copy.cpp

```diff
@@ -0,0 +1,150 @@
+}
+
+template <typename T>
+typename std::enable_if<!is_vec<T>::value, bool>::type checkEqual(T A,
+                                                                  size_t B) {
+  T TB = B;
+  return A == TB;
+}
+
+template <typename T> std::string toString(vec<T, 1> A) {
+  std::string R("(");
+  return R + std::to_string(A.s0()) + ")";
+}
+
+template <typename T> std::string toString(vec<T, 4> A) {
+  std::string R("(");
+  R += std::to_string(A.x()) + "," + std::to_string(A.y()) + "," +
+       std::to_string(A.z()) + "," + std::to_string(A.w()) + ")";
+  return R;
+}
+
+template <typename T = void>
+typename std::enable_if<!is_vec<T>::value, std::string>::type toString(T A) {
+  return std::to_string(A);
+}
+
+template <typename T> int checkResults(buffer<T, 1> &OutBuf, size_t Stride) {
+  auto Out = OutBuf.template get_access<access::mode::read>();
+  int EarlyFailout = 20;
+
+  for (size_t I = 0; I < OutBuf.get_count(); I += WorkGroupSize) {
+    for (size_t J = 0; J < WorkGroupSize; J++) {
+      size_t ExpectedVal = (J % Stride == 0) ? (100 + I + J) : 0;
+      if (!checkEqual(Out[I + J], ExpectedVal)) {
+        std::cerr << std::string(typeid(T).name()) +
+                         ": Incorrect value at index "
+                  << I + J << " : "
+                  << "Expected: " << toString(ExpectedVal)
+                  << ", Computed: " << toString(Out[I + J]) << "\n";
+        if (--EarlyFailout == 0)
+          return 1;
+      }
+    }
+  }
+  return EarlyFailout - 20;
+}
+
+template <typename T> int test(size_t Stride) {
+  queue Q;
+
+  buffer<T, 1> InBuf(NElems);
+  buffer<T, 1> OutBuf(NElems);
+
+  initInputBuffer(InBuf, Stride);
+  initOutputBuffer(OutBuf);
+
+  Q.submit([&](handler &CGH) {
+     auto In = InBuf.template get_access<access::mode::read>(CGH);
+     auto Out = OutBuf.template get_access<access::mode::write>(CGH);
+     accessor<T, 1, access::mode::read_write, access::target::local> Local(
+         range<1>{WorkGroupSize}, CGH);
+
+     nd_range<1> NDR{range<1>(NElems), range<1>(WorkGroupSize)};
+     CGH.parallel_for<KernelName<T>>(NDR, [=](nd_item<1> NDId) {
+       auto GrId = NDId.get_group_linear_id();
+       auto Group = NDId.get_group();
+       size_t NElemsToCopy =
+           WorkGroupSize / Stride + ((WorkGroupSize % Stride) ? 1 : 0);
+       size_t Offset = GrId * WorkGroupSize;
+       auto E = NDId.async_work_group_copy(Local.get_pointer(),
+                                           In.get_pointer() + Offset,
+                                           NElemsToCopy, Stride);
+       E.wait();
+
+       E = NDId.async_work_group_copy(Out.get_pointer() + Offset,
+                                      Local.get_pointer(), NElemsToCopy,
+                                      Stride);
+       Group.wait_for(E);
+     });
+   }).wait();
+
+  return checkResults(OutBuf, Stride);
+}
+
+int main() {
+  for (int Stride = 1; Stride < WorkGroupSize; Stride++) {
```

Comment:
Are there any possible negative scenarios?

Reply:
I'd rather keep this test as is now without additional changes in initialization/correctness-checks, if you don't mind.

Reply:
BTW, by "negative scenarios", did you mean negative Stride?
Currently, SYCL SPEC specifies the Stride arg as unsigned, i.e. size_t
59 template <typename dataT>
60 device_event async_work_group_copy(local_ptr<dataT> dest,
61 global_ptr<dataT> src, size_t numElements, size_t srcStride) const;
62

Reply:
> BTW, by "negative scenarios", did you mean negative Stride?

Not really. "Negative" scenario is the one that should result negative or with failure i.e. when copy operation isn't allowed or shouldn't succeed.